Add SSL/TLS support

pemontto opened the pull request:
This PR adds the ability to configure password protected key and certificate, or keystore and truststore transport and HTTP encryption.

If both are defined it will preference keystores and truststore based on the preference in the Elasticsearch documentation. It doesn't yet support having different keys for HTTP and trasport, though that won't be hard to add.

Currently a WIP because it's not documented and I have yet to define tests to cover various use cases. I'm having trouble getting converge running, it seems to pick up some strange paths when "preparing roles".

The role attempts to add or remove passwords to/from the keystore only when the cert/truststore copy operation is changed. This might be a non-issue but it could lead to an edge case where if alternate key/keystores have been uploaded already, swapping between them won't update the password to unlock them in the keystore.

```shell
Failed to complete #converge action: [No such file or directory @ rb_sysopen - /Users/test/git/<some other git repo>/.venv/.Python] on oss-centos-7
```

elasticmachine commented:
Since this is a community submitted pull request, a Jenkins build has not been kicked off automatically. Can an Elastic organization member please verify the contents of this patch and then kick off a build manually?

pemontto commented:
This builds on the work of @AevaOnline from #597 and @kcm from #479

pemontto commented:
@jmlrt this is ready to look at, are you able to kick off a jenkins build?

I had to do my testing with the new trial task to test security under version 7.1.0.

TODO: documentation

jmlrt commented:
jenkins test this please

jmlrt commented:
jenkins test this please

pemontto commented:
Documentation added.

This PR now addresses #621, #597, #479, #500 and relates to #586

jmlrt commented:
jenkins test this please

pemontto commented:
@jmlrt updated test suite and merged in latest master branch changes

jmlrt commented:
@pemontto, thanks for this amazing PR, I still didn't find the time to start the review outside of launching regular jenkins tests. I hope being able to review it last week.

jmlrt commented:
Jenkins test this please

jmlrt commented:
jenkins test this please

cstegm commented on tasks/main.yml:
```diff
@@ -76,18 +82,22 @@
-- name: Wait 15 seconds for the Native Relm to come up
-  command: sleep 15
-  when: manage_native_realm
+- name: Wait {{ es_api_sleep }} seconds for the Native Realm to come up
```
Why not just let the user override it? I Bad issues with this hardcoded value. #485

pemontto commented on README.md:
```diff
@@ -69,6 +70,10 @@ This playbook uses [Kitchen](https://kitchen.ci/) for CI and local testing.
 
+* Ensure you have checked out this repository to `elaticsearch`, not `ansible-elasticsearch`.
+* If you don't have a Gold or Platinum license to test with you can run the trial versions of the `xpack-upgrade` and `issue-test` suites by appending `-trial` to the `PATTERN` variable.
+* You may need to explicity specify `VERSION=7.x` if some suites are failing.
```
Yeah I assumed that would default to 7.x, but for some suites if it's not explicitly defined in the command it fails. This doesn't show up in Jenkins because it always specifies VERSION=x

E.g. `make verify PATTERN=oss-upgrade-centos-7`

```shell
       TASK [elasticsearch : set_fact] ************************************************
       Thursday 31 October 2019  09:45:14 +0000 (0:00:08.938)       0:00:09.084 ******
       fatal: [localhost]: FAILED! => {"msg": "The task includes an option with an undefined variable. The error was: 'es_major_version' is undefined\n\nThe error appears to be in '/tmp/kitchen/elasticsearch/tasks/main.yml': line 3, column 3, but may\nbe elsewhere in the file depending on the exact syntax problem.\n\nThe offending line appears to be:\n\n\n- set_fact: \"es_major_version={{ es_version.split('.')[0] }}.x\"\n  ^ here\nWe could be wrong, but this one looks like it might be an issue with\nmissing quotes. Always quote template expression brackets when they\nstart a value. For instance:\n\n    with_items:\n      - {{ foo }}\n\nShould be written as:\n\n    with_items:\n      - \"{{ foo }}\"\n"}

       PLAY RECAP *********************************************************************
       localhost                  : ok=1    changed=0    unreachable=0    failed=1    skipped=0    rescued=0    ignored=0

       Thursday 31 October 2019  09:45:14 +0000 (0:00:00.025)       0:00:09.110 ******
       ===============================================================================
       Gathering Facts --------------------------------------------------------- 8.94s
       elasticsearch : set_fact ------------------------------------------------ 0.03s
>>>>>> ------Exception-------
>>>>>> Class: Kitchen::ActionFailed
>>>>>> Message: 1 actions failed.
>>>>>>     Converge failed on instance <oss-upgrade-centos-7>.  Please see .kitchen/logs/oss-upgrade-centos-7.log for more details
>>>>>> ----------------------
>>>>>> Please see .kitchen/logs/kitchen.log for more details
>>>>>> Also try running `kitchen diagnose --all` for configuration
```

jmlrt replied:
OK it seems that `VERSION := 7.x` in the `Makefile` isn't forwarded to kitchen as environment variable, however if you override that variable with `make VERSION=7.x`, it is forwarded to kitchen.

I'll work on a fix for that.

pemontto commented on README.md:
```diff
@@ -69,6 +70,10 @@ This playbook uses [Kitchen](https://kitchen.ci/) for CI and local testing.
 
 ### Running the tests
 
+* Ensure you have checked out this repository to `elaticsearch`, not `ansible-elasticsearch`.
```
And I still managed to make a type 🤦‍♂

pemontto commented on ansible.cfg:
```diff
@@ -1 +1,2 @@
-[defaults]
\ No newline at end of file
+[defaults]
+callback_whitelist = profile_tasks
```
Happy to remove it, I thought it could be useful for historical comparison, and I believe it's only use when the test suite is run. It won't be used by anyone consuming the role right?

jmlrt replied:
Yeah also it's really useful, we don't currently use it and I would prefer removing it of the repository even if there isn't any impact for users consuming the role.

pemontto replied:
Removed!

pemontto commented:
@jmlrt I've moved the cert files along with the template files and updated the integration tests:
```
files/certs -> test/integration/files/certs
files/templates-6.x -> test/integration/files/templates-6.x
files/templates-7.x -> test/integration/files/templates-7.x
```

jmlrt commented:
jenkins test this please

jmlrt commented:
>@jmlrt I've moved the cert files along with the template files and updated the integration tests:

Nice catch for the templates 👍

pemontto commented:
devops failure looks to be a transient [debian-8 ](https://devops-ci.elastic.co/job/elastic+ansible-elasticsearch+pull-request/99/OS=debian-8,TEST_TYPE=oss-to-xpack-upgrade,VERSION=7.x,label=linux/console) issue

jmlrt commented:
>devops failure looks to be a transient debian-8 issue

Yes that's not related to your changes so not a blocker.

Local tests are OK for me.

I would prefer removing the `callback_whitelist` of the role.

I will also ask for a second opinion from elastic people, especially for the idea of generating tls configuration in `elasticsearch.yml.j2` versus letting users add it in `es_config`.

jmlrt commented:
jenkins test this please

Crazybus commented on templates/elasticsearch.yml.j2:
```diff
@@ -33,6 +33,39 @@ action.auto_create_index: {{ es_action_auto_create_index }}
 {% if es_enable_xpack and es_api_basic_auth_username is defined and es_api_basic_auth_password is defined %}
 xpack.security.enabled: true
+
+{% if es_enable_transport_ssl | bool %}
```
I think its ok to leave these autogenerated values in as long as there is a way to opt out of it by adding some kind of `es_enable_auto_ssl_configuration: false`. Right now if someone wanted to add multiple `certificate_authorities` (which is an array/list) they wouldn't be able to override it in `es_config` because those values are being provided first and assume that you only have one CA. There are probably a bunch of other edge cases that could prevent a user from enabling security if they were forced to have these exact values set.

pemontto replied:
Good point, will add a `es_enable_auto_ssl_configuration` parameter and test.

jmlrt commented:
jenkins test this please

jmlrt commented:
One failing test not related to PR code (timeout during ES package install).

Let's merge this PR!

@pemontto but also @AevaOnline (#597), @itsmed & @kcm (#479), @strootman (#302), @xocasdashdash (#635), @justbkuz (#442), @cdahlqvist (#331), thanks you all for your amazing work to make this Ansible role supporting with TLS 👍 👍 👍
